### ISSUE TITLE
Command buffers

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4406]
Command buffers<br>
<br>
http://bit.ly/2MQFt2A<br>
<br>
About 0 Minutes<br>
<br>
<div>
<section>
<img width="100%" src="http://alinloghin.com/articles/command_buffer/front.png" height="100%">
</section> <section>
<div>
<div>
<div>
<h2>Command Buffer</h2>
<p>In this post I will present a C++ lock-free and API agnostic command buffer implementation that can be used for multi-threaded rendering and reducing overhead (state changes, depth sorting) on the main thread. The implementation is based on Stefan Reinalter's blog post<sup>1</sup> on multi-thread rendering which I recommend reading as it shows several alternatives and more insight to certain aspects. It also includes a configurable draw key class (via bitfields), implementation of some basic GL commands and supports multiple material passes. </p>
<p>In most cases you will have multiple command buffers for handling specific rendering logic like in the case of a deferred renderer: shadow maps, g-buffer pass, deferred pass and a post-process pass. <br> Commands will be dispatched into one or more of the command buffers from multiple threads (main thread can also participate as it shouldn't be idle waiting), after which you'll have to wait until all commands are dispatched and it's safe to sort and submit the commands (to the graphics driver) on the main thread:</p>
<center><img width="80%" src="http://alinloghin.com/articles/command_buffer/cmd_diagram.png" height="80%"></center>
<p>Note that in some cases certain command buffers dont require frequent updates (for example the post processing command buffer) as they only need to be updated when something changes. <br>Defining a command buffer requires a key type with its decoder class and an optional material binder:</p>

<div>
<pre>    <span>typedef</span> cb<span>::</span>CommandBuffer<span>&lt;</span>cb<span>::</span>DrawKey, cb<span>::</span>DefaultKeyDecoder, Nv<span>::</span>MaterialBinder<span>&gt;</span> GeometryCommandBuffer;
    <span>typedef</span> cb<span>::</span>CommandBuffer<span>&lt;</span><span>uint32_t</span>,cb<span>::</span>DummyKeyDecoder<span>&lt;</span><span>uint32_t</span><span>&gt;&gt;</span> DeferredCommandBuffer;
    <span>typedef</span> cb<span>::</span>CommandBuffer<span>&lt;</span><span>uint16_t</span>,cb<span>::</span>DummyKeyDecoder<span>&lt;</span><span>uint16_t</span><span>&gt;&gt;</span> PostProcessCommandBuffer;
</pre>
</div>
<br>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Command Key</h3>
<p>The command key will be used for sorting the commands, its layout really depends on the use case of the command buffer, for example a shadow command buffer would simply use a 32 bit command representing the depth while for the g-buffer it may be a 64 bit integer that is structured like<sup>2</sup>:</p>
<center><img width="75%" src="http://alinloghin.com/articles/command_buffer/cmd_key.png" height="75%"></center>
<p>At the most significant bits there is a viewport ID (for handling multiple viewports), then a view layer (e.g. skybox, world, etc.), translucency type (e.g. opaque, normal, additive, subtractive), a command bit and the rest of the bits will be shared by 3 commands modes: opaque, transparent and custom.</p>
<p>For the transparent and opaque modes the depth and material ID fields are swapped and a default key decoder handles this:</p>

<div>
<pre>    <span>struct</span> DefaultKeyDecoder
    {
      <span>/// note Each command key must define a decode function that will be used by the command buffer.</span>
      CB_FORCE_INLINE cb<span>::</span>MaterialId <span>operator</span>()(cb<span>::</span>DrawKey key) <span>const</span>
      {
        <span>return</span> key.material();
      }
    };
    <span>// decode function, get correct material ID as it is swapped</span>
    <span>inline</span> cb<span>::</span>MaterialId DrawKey<span>::</span>material() <span>const</span>
    {
      <span>return</span> isOpaqueMode() <span>?</span> cb<span>::</span>MaterialId(opaque.materialId, opaque.materialPass)
        <span>:</span> cb<span>::</span>MaterialId(transparent.materialId, transparent.materialPass);
    }
</pre>
</div>
<p>The reason we want to do front to back ordering for the opaque mode is to take advantage of hardware early z-culling, if we're doing a depth pre-pass like most engines do then this can be ignored for opaque objects, this will discard occluded fragments: </p>
<center><img width="75%" src="http://alinloghin.com/articles/command_buffer/depth_order.png" height="75%"></center>
<p>The lower bits of the material ID are used by the pass number and can be used to handle multiple material subpasses:</p>

<div>
<pre>      <span>bool</span> nextPass;
      <span>do</span>
      {
        <span>// if true then we have more passes presents in the material</span>
        nextPass <span>=</span> m_materialBinder(material);
<span>#if CB_DEBUG_COMMANDS_PRINT</span>
        m_materialBinder.debugMsg(material);
        CommandPacket<span>::</span>log(packet, <span>*</span>m_logger);
<span>#endif</span>
        CommandPacket<span>::</span>dispatch(packet, rc);
        <span>++</span>material.pass;
      } <span>while</span> (nextPass);
</pre>
</div>
<p>The are multiple ways of creating a command key:</p>

<div>
<pre>    cb<span>::</span>DrawKey key(<span>0</span>);
    key.setViewLayer(cb<span>::</span>ViewLayerType<span>::</span>e3D, cb<span>::</span>TranslucencyType<span>::</span>eOpaque);
    key.setDepth(depth); <span>// handles swapping based on translucency type </span>
    key.setMaterial(materialId);
    <span>// using make </span>
    key <span>=</span> cb<span>::</span>DrawKey<span>::</span>makeDefault(viewportId, cb<span>::</span>ViewLayerType<span>::</span>e3D);
    key <span>=</span> cb<span>::</span>DrawKey<span>::</span>makeCustom(viewportId, cb<span>::</span>ViewLayerType<span>::</span>e3D, priorityNumber);
</pre>
</div>
<p>Finally, custom commands are enabled via the command bit, they can be used for betweeen layers (e.g. view layers, viewports) and can be ordered via a priority (or via command chaining), this way you can change the render target or update vertex buffers before the draw calls.</p>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Creating Commands</h3>
<p>Creating commands is straightforward, it requires creating a POD struct and defining its dispatch function:</p>

<div>
<pre>    <span>struct</span> DrawArrays {
        <span>static</span> <span>const</span> cb<span>::</span>RenderContext<span>::</span><span>function_t</span> kDispatchFunction;
        GLuint          vao;
        <span>uint32_t</span>        base;
        <span>uint32_t</span>        count;
        GLenum          primitive;
    };
    <span>//in cpp</span>
    <span>void</span> <span>drawArrays</span>(<span>const</span> <span>void</span><span>*</span> data, cb<span>::</span>RenderContext<span>*</span>) {
        <span>auto</span><span>&amp;</span> cmd <span>=</span> <span>*</span><span>reinterpret_cast</span><span>&lt;</span><span>const</span> DrawGroundCommand<span>*&gt;</span>(data);
        glBindVertexArray(cmd.vao);
        glDrawArrays(cmd.primitive, cmd.base, cmd.count);
    }
    <span>const</span> cb<span>::</span>RenderContext<span>::</span><span>function_t</span> DrawArrays<span>::</span>kDispatchFunction <span>=</span> <span>&amp;</span>drawArrays;
</pre>
</div>
<p>All commands must be PODs as the default constructor or destructor will not be called due to memory allocation logic, otherwise there will be a compile error, however in cases where you have members that have contiguous memory (or want to handle initialization/allocation manually) but aren't PODs (e.g. a matrix class with constructors) you can hint that the structure is actually a POD via:</p>

<div>
<pre>    <span>struct</span> DrawPointLightCommand
    {
        <span>static</span> <span>const</span> cb<span>::</span>RenderContext<span>::</span><span>function_t</span> kDispatchFunction;
        <span>// hint that we dont care about ctr/dtr</span>
        <span>typedef</span> <span>void</span> pod_hint_tag;
        
        LightingUBO lightingUBO_Data; <span>// non-pod member</span>
</pre>
</div>
<p>For setting up the dispatch function with a member method use makeExecuteFunction that will use the execute member method of the command class:</p>

<div>
<pre>    <span>struct</span> RenderNonInstanced
    {
      <span>static</span> <span>const</span> cb<span>::</span>RenderContext<span>::</span><span>function_t</span> kDispatchFunction;

      NvModelExtGL<span>*</span> pSourceModel;
      GLint positionHandle, normalHandle, texcoordHandle, tangentHandle;

      <span>void</span> <span>execute</span>() <span>const</span>
      {
        pSourceModel<span>-&gt;</span>DrawElements(<span>1</span>, positionHandle, normalHandle, texcoordHandle, tangentHandle);
      }
    };
   <span>//in cpp</span>
   <span>const</span> cb<span>::</span>RenderContext<span>::</span><span>function_t</span>   RenderNonInstanced<span>::</span>kDispatchFunction <span>=</span> <span>&amp;</span>cb<span>::</span>makeExecuteFunction<span>&lt;</span>RenderNonInstanced<span>&gt;</span>;
</pre>
</div>
<p>Finally note that all the commands must match the alignment of the command buffer allocator. You can match the aligment using the CB_COMMAND_PACKET_ALIGN macro anywhere inside the struct definition or you can disable alignment requirements by setting CB_COMMAND_PACKET_ALIGNED macro to zero (this will result in slightly lower memory consumption of the commands).</p>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Material Binder</h3>
<p>In order to reduce state changes, such as binding the same shader or updating uniform buffers with redundant data, a material binder can be used which will handle the logic of setting the active material, note that commands are sorted according to their material id. <br>For example a material can be defined as:</p>

<div><pre>    <span>struct</span> Material
    {
        std<span>::</span>string name;
        GLuint location <span>=</span> <span>0</span>;
        GLuint ubo <span>=</span> <span>0</span>;
        NvGLSLProgram<span>*</span> shader <span>=</span> nullptr;
    };
</pre></div>
<p>Afterwards you have to implement the bind operator of the material binder class, for example:</p>

<div><pre>    <span>// @note Returns true if there are more passes to bind.</span>
    CB_FORCE_INLINE <span>bool</span> <span>operator</span>()(cb<span>::</span>MaterialId material) <span>const</span>
    {
        <span>if</span> (material.id <span>==</span> <span>0</span>) <span>// first material is a dummy, nothing to bind</span>
            <span>return</span> <span>false</span>;

        <span>const</span> Material<span>&amp;</span>  mat <span>=</span> materials[material.id];
        assert(mat.shader);
        <span>if</span> (mat.shader <span>!=</span> activeShader)
        {
            <span>//bind only if different shaders</span>
            mat.shader<span>-&gt;</span>enable();
            activeShader <span>=</span> mat.shader;
        }
        <span>if</span> (material.id <span>!=</span> activeMaterial)
        {
            <span>// bind material ubo</span>
            glBindBufferBase(GL_UNIFORM_BUFFER, mat.location, mat.ubo);
            activeMaterial <span>=</span> material.id;
        }

        <span>// @note you can use material.pass to handle multiple passes for the same material</span>

        <span>return</span> <span>false</span>;
    }
</pre></div>
<p>You can also handle multiple material subpasses by using material.pass. The material binder is optional and by default the command buffer will use a dummy binder, i.e.: cb::DefaultMaterialBinder.</p>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Dispatching Commands</h3>
<p>There are multiple way to dispatch commands and the most common is using addCommand, each time you'll have to use a command key (if possible it would best to save and reuse it):</p>

<div>
<pre>    <span>//create the command key first</span>
    cb<span>::</span>DrawKey key(<span>0</span>);
    key.setViewLayer(cb<span>::</span>ViewLayerType<span>::</span>e3D, cb<span>::</span>TranslucencyType<span>::</span>eOpaque);
    key.setDepth(depth);
    key.setMaterial(materialId);
    
    <span>//then create command which will dispatch it</span>
    cmds<span>::</span>DrawArrays<span>&amp;</span> cmd <span>=</span> <span>*</span>commandBuffer.addCommand<span>&lt;</span>cmds<span>::</span>DrawArrays<span>&gt;</span>(key);
    cmd.vao <span>=</span> vaoId;
    cmd.base <span>=</span> <span>0</span>;
    cmd.count <span>=</span> <span>4</span>;
    cmd.primitive <span>=</span> GL_TRIANGLE_STRIP;
</pre>
</div>
<p>Multiple commands can also be chained together by using appendCommand, that's if they can share the same command key (material, depth, etc.) which will avoid the overhead of material binder:</p>

<div>
<pre>    cb<span>::</span>DrawKey key <span>=</span> cb<span>::</span>DrawKey<span>::</span>makeDefault(viewportId, cb<span>::</span>ViewLayerType<span>::</span>e3D);
    <span>//setup key</span>
    cmds<span>::</span>DrawArrays<span>*</span> cmd <span>=</span> commandBuffer.addCommand<span>&lt;</span>cmds<span>::</span>DrawArrays<span>&gt;</span>(key);
    <span>//fill command data</span>
    CB_DEBUG_COMMAND_SET_MSG(cmd, <span>"draw first quad"</span>);
    
    <span>//append multiple draw calls</span>
    <span>for</span>(<span>int</span> i <span>=</span> <span>1</span>; i <span>&lt;</span> <span>10</span>; <span>++</span>i) {
        cmd <span>=</span> commandBuffer.appendCommand<span>&lt;</span>cmds<span>::</span>DrawArrays<span>&gt;</span>(cmd);
        <span>//fill command data</span>
    ...
</pre>
</div>
<p>Commands can also have auxiliary data, like in the case of updating an uniform buffer:</p>

<div>
<pre>    std<span>::</span>vector<span>&lt;</span><span>float</span><span>&gt;</span> data;
    <span>// create command and allocate auxiliary data</span>
    cmds<span>::</span>UpdateUniformCommand<span>&amp;</span> cmd <span>=</span> <span>*</span>commandBuffer.addCommand<span>&lt;</span>cmds<span>::</span>UpdateUniformCommand<span>&gt;</span>(key, <span>sizeof</span>(<span>float</span>) <span>*</span> data.size());
    <span>// get the command packet from the command and manually copy the data</span>
    CommandPacket<span>*</span> packet <span>=</span> CommandPacket<span>::</span>getCommandPacket<span>&lt;</span>CommandClass<span>&gt;</span>(cmd);
    memcpy(packet<span>-&gt;</span>auxilaryData, data.data(), <span>sizeof</span>(<span>float</span>) <span>*</span> data.size());
    cmd.data = packet<span>-&gt;</span>auxilaryData;
    cmd.size = sizeof(float) * data.size();
    <span>// or you can use the template version which will automatically do the copy</span>
    cmds<span>::</span>UpdateUniformCommand<span>&amp;</span> cmd <span>=</span> <span>*</span>commandBuffer.addCommandData<span>&lt;</span>cmds<span>::</span>UpdateUniformCommand<span>&gt;</span>(key, data);
    
    <span>// this will set the data and size members with packet-&gt;auxilaryData and size in bytes</span>
    <span>struct</span> UpdateUniformCommand
    {
        <span>static</span> <span>const</span> tn<span>::</span>gpu_dispatch<span>::</span><span>function_t</span> kDispatchFunction;
        <span>// the command must have a data and size as members</span>
        <span>const</span> <span>void</span><span>*</span> data;
        <span>uint32_t</span> size;
        ...
</pre>
</div>
<p>Multiple commands can also share a command (i.e. header packet), for example in a case where you might want to avoid the overhead of copying the command's data:</p>

<div>
<pre>    <span>//create the first the shared command</span>
    CommandPacket<span>*</span> headerPacket <span>=</span> buffer.createCommandPacketData<span>&lt;</span>cmds<span>:</span>SetMatrixCommand<span>&gt;</span>(matrices);

    <span>//reference/share the command</span>
    key <span>=</span> cb<span>::</span>DrawKey<span>::</span>makeDefault(viewportId, cb<span>::</span>ViewLayerType<span>::</span>e3D);
    CommandPacket<span>*</span> cmd <span>=</span> buffer.addCommandFrom(key, headerPacket);
    <span>//append a new command to the shared one</span>
    cmds<span>::</span>DrawArrays<span>*</span> draw <span>=</span> buffer.appendCommand<span>&lt;</span>cmds<span>::</span>DrawArrays<span>&gt;</span>(cmd);
    <span>//fill command data</span>

    <span>//reference it again</span>
    key <span>=</span> cb<span>::</span>DrawKey<span>::</span>makeDefault(otherViewportId, cb<span>::</span>ViewLayerType<span>::</span>e3D);
    cmd <span>=</span> buffer.addCommandFrom(key, headerPacket);
    draw <span>=</span> buffer.appendCommand<span>&lt;</span>cmds<span>::</span>DrawArrays<span>&gt;</span>(cmd);
    <span>//fill command data</span>
    ...
</pre>
</div>
<br>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Submiting Commands</h3>
<p>Before we can submit the commmands to the GPU we first have to sort them (this can also be done on multiple threads if we have too many commands), a custom sort function can also be used:</p>

<div>
<pre>    m_geometryCommands.sort(radixsort<span>&lt;</span>GeometryCommandBuffer<span>::</span><span>command_t</span><span>&gt;</span>);
    m_deferredCommands.sort();
    m_postProcessCommands.sort();
</pre>
</div>
<p>Note that above we used radix sort for the geometry commands, this is due to the way the command key is defined a radix sort will be a little faster in most cases. <br> Afterwards we simply have to call submit (note that we can pass an optional context data like in the case of D3D):</p>

<div><pre>    <span>const</span> <span>bool</span> clearCommands <span>=</span> <span>!</span>m_animPaused; <span>// using recorded commands when paused</span>
    <span>// nothing to pass as GL doesn't have any context data</span>
    m_geometryCommands.submit(nullptr, clearCommands);
    m_deferredCommands.submit(nullptr, clearCommands);
    m_postProcessCommands.submit(nullptr, clearCommands);
</pre></div>
<p>Due to the fact we copy in the commands all the data required for rendering when dispatching a command, we can reuse the submitted commands and not clear the command buffers. However to be optimal, we must make sure we don't update any resources (like vertex buffers or textures) in those commands, depending on your use case you could use separate command buffers only for handling resource management.</p>
<p>This can be used to reduce overhead in certain cases, but can also be useful for debugging purposses as you could dump the commands to disk and replay them later. As long you can serialize the commands, in the case of using raw pointers in the commands you have to hard copy the data, and the asset's data (i.e. textures, geometry).</p>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Debugging</h3>
<p>One of the disadvantages of using command buffers compared to a stateful approach is that in some cases it makes debbuging a bit more difficult, this is due to the fact that commands are deferred and submitted on the main thread.</p>
<p>To deal with this tagging commands can be enabled via the CB_DEBUG_TAG_COMMANDS macro and aftewards using:</p>

<div>
<pre>    cb<span>::</span>DrawKey key <span>=</span> cb<span>::</span>DrawKey<span>::</span>makeDefault(viewportId, cb<span>::</span>ViewLayerType<span>::</span>e3D);
    cmds<span>::</span>DrawArrays<span>&amp;</span> cmd <span>=</span> <span>*</span>commandBuffer.addCommand<span>&lt;</span>cmds<span>::</span>DrawArrays<span>&gt;</span>(key);
    ...
    <span>// will name tag: 'testFunction : &lt;cpp_line_number&gt; : struct cmds::DrawArrays * __ptr64'</span>
    CB_DEBUG_COMMAND_TAG(cmd);
    <span>// will name tag: 'testFunction : &lt;cpp_line_number&gt; : struct cmds::DrawArrays * __ptr64 : draw quad'</span>
    CB_DEBUG_COMMAND_TAG_MSG(cmd, <span>"draw quad"</span>);
    <span>// will name tag: 'draw quad'</span>
    CB_DEBUG_COMMAND_SET_MSG(cmd, <span>"draw quad"</span>);
    <span>// can also use a compile time sprint</span>
    CB_DEBUG_COMMAND_SET_MSG(cmd, FORMAT_STR(<span>"draw quad %d"</span>, index));
</pre>
</div>
<p>Enabling command tagging will add packet_debug member inside the CommandPacket:</p>

<div>
<pre>    <span>struct</span> packet_debug
    {
        <span>static</span> <span>const</span> <span>int</span> kSize <span>=</span> <span>128</span>;
        <span>char</span> tag[kSize];
    };
</pre>
</div>
<p>You can also CB_DEBUG_COMMANDS_PRINT macro and if needed define a custom log function (by default it'll use stdout):</p>

<div>
<pre>    <span>//custom log function</span>
    <span>int</span> <span>commandLogFunction</span>(<span>const</span> <span>char</span><span>*</span> fmt, ...)
    {
      <span>va_list</span> args;
      <span>char</span> buffer[1024];

      va_start(args, fmt);
      vsnprintf(buffer, <span>sizeof</span> buffer, fmt, args);
      va_end(args);

      NVPlatformLog(buffer);
      <span>return</span> <span>0</span>;
    }
    <span>//set the log function after</span>
    m_geometryCommands.setLogFunction(<span>&amp;</span>commandLogFunction);
</pre>
</div>
<br>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Final Thoughts</h3>
<p>In the newer graphics APIs, like Vulkan and D3D12, the concept of command buffer has been introduced at the API level. They can be used with multiple threads (you can have secondary buffers) or be recorded, still this implementation can be useful with new APIs due to its other features such as: sorting via command keys, API agnostic, material binder and multiple passes, it's also less verbose to use.</p>
<p>However certain changes may be required, for example the thread-safe logic (like the lock-free linear allocator) could be an extra template argument only used in cases when it's required to be thread-safe, thus only if you dispatch commands on multiple threads and submit on one.</p>
</div>
</div>
</div>
</section>
<section>
</section>
<hr noshade>

<span></span>
</div><br>
<br>
<br>
<br>
<br>
<br>
via alinloghin.com http://alinloghin.com<br>
<br>
June 18, 2019 at 02:25PM
